src/controller/worker.rs

```diff
@@ -0,0 +1,161 @@
+use std::sync::Arc;
+
+use futures::{Future, StreamExt};
+use subsquid_network_transport::PeerId;
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    gateway_allocations::{self, allocations_checker::AllocationsChecker},
+    metrics,
+    query::{self, error::QueryError, eth::BatchRequest, result::QueryResult},
+    storage::{
+        datasets_index::DatasetsIndex,
+        manager::{self, StateManager},
+    },
+    types::{dataset::Dataset, state::ChunkSet},
+    util::UseOnce,
+};
+
+lazy_static::lazy_static! {
+    static ref PARALLEL_QUERIES: usize = std::env::var("PARALLEL_QUERIES")
+        .map(|s| s.parse().expect("Invalid PARALLEL_QUERIES"))
+        .unwrap_or(3);
+    static ref QUEUED_QUERIES: usize = std::env::var("QUEUED_QUERIES")
+        .map(|s| s.parse().expect("Invalid QUEUED_QUERIES"))
+        .unwrap_or(5);
```

Comment:
5 queued queries seems a bit low to me. What was the logic behind choosing this number?

Reply:
I calculated like this: if the average time for query execution is ~200ms and we can tolerate a delay of 1000ms, we get only 5 queries queued. But I forgot to multiply by the number of parallel queries, so it should probably be 15.

Reply:
Yeah, 15 sounds better.